### ISSUE TITLE
Coloring zones by source location collapses colors incorrectly when zoomed out.

### ISSUE THREAD

**Issue by mcourteaux** [wolfpld/tracy#536]
When coloring zones by source code location, colors are inconsistent throughout zooming.
When zooming out, the individual zones change color. Also the collapsed zones (with the zigzag pattern) adopt a wrong color.

https://user-images.githubusercontent.com/845012/222916433-f29c3721-c9d6-4a4c-9407-f119a05d20de.mp4

**Comment by wolfpld** [comment]
This is working as intended. The alternative would be to use the color of first collapsed zone, which would result in a rave party when you're panning the view and differently colored zones happen to be first. It is not possible to determine the color of all collapsed zones in a performant way.

**Comment by mcourteaux** [comment]
Okay, so in principle, the color should be "undetermined", as taking the mean over the group of collapsed Zones is a bottleneck. But then why is the collapsed color still variable from thread to thread? Is this because the "undetermined" color falls back to using the per-thread coloring scheme?

**Comment by mcourteaux** [comment]
The reason I'm asking is, because I thought that the reason of the per-source-location coloring scheme was made to easily spot the function-calling patterns throughout time. This per-thread-fallback color for collapsed zones makes this counterintuitive, as it suggests it's not the same Zones being represented.

**Comment by wolfpld** [comment]
The fallback color is the thread color. What color would you like to see instead?

**Comment by mcourteaux** [comment]
I think it should be possible give a better color in a performant way. I see a few options (which are all O(1) at display time):

 - make it simply gray to reflect the fact that it's unknown.
 - just pick a desaturated variant of the parent zone
 - when child/parent relationships are resolved/processed, average out the colors of the child, and use an average color (disregarding the fact that not all childs are responsible for a collapsed zone, so less favorable).
 - a little more elaborate, but you could maintain a cumulative color function.

Let me elaborate on the last option. This is a common technique to speed up taking arbitrary sums (and hence averages) of subregions of static data. A collapsed region is basically:

```cpp
collapsed_region_color = 1/(#zones) * sum[z=zones](z.color)
```

Pretty similar to finding a primitive function of a function using an undetermined integral in calculus, you can do the same for a discrete summation like the one above. Specifically, each zone would have a `cumulative_zone_color` being equal to all the sum of all the colors of it's earlier (ie: towards the left on the timeline) siblings (earlier zones under the same parent zone). So you basically store the primitive function in the zones.

Then when you collapse a region of zones from zone `i` to zone `j`, you can quickly recover the sum of all the zones in this collapsed region by taking:

```cpp
sum_color = zones[j+1].cumulative_zone_color - zones[i].cumulative_zone_color
// for r, g, b separately of course
```

Now all is left is dividing (to go from the sum to the mean), and checking boundary conditions etc (regarding the +1).

As I said, somewhat more elaborate, and I don't know the architecture of how the zones are internally managed and stored, but this seems like it could be the basis of a performant solution.

**Comment by mcourteaux** [comment]
Of course, I don't care that much about these colors. This is such a minor thing. I just want to express how great Tracy is once again. This tool really makes me a better engineer and programmer. Thanks! <3

**Comment by wolfpld** [comment]
Fixed by going gray in 107afd78.

**Comment by mcourteaux** [comment]
Cool! Thanks :smile:
